bluesky/run_engine.py

```diff
@@ -1353,7 +1353,18 @@ def _open_run(self, msg):
         self._run_start_uids.append(self._run_start_uid)
 
         # Run scan_id calculation method
-        self.md['scan_id'] = self.scan_id_source(self.md)
+        stateful_md = self.md.copy()
+        if self.scan_id_source is not None:
+            scan_id = self.scan_id_source(stateful_md)
+        else:
+            scan_id = None
+
+        if scan_id is None:
+            if 'scan_id' in stateful_md:
```

Comment:
you can use `stateful_md.pop('scan_id', None)` which will do the same thing with one less conditional.